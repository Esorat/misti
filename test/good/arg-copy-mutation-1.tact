contract Test {
    m: map<Int, Int>;
    fun test(a: Int, m: map<Int, Int>) {
        // Bad cases:
        m.set(3, 3);
        m.del(3);

        // OK: No warning since nobody will expect an integral type to be modified
        a += 1;
        a = 42;

        // OK: Accessing map element
        let _b: Int? = m.get(3);

        // OK: Modifying contract fields
        self.m.set(3, 3);
    }

    // To ensure we report the argument once per function
    fun testOnce(m: map<Int, Int>) {
        m.set(1, 1);
        m.set(2, 2);
    }
}
