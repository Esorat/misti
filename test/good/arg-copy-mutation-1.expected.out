[HIGH] ArgCopyMutation: Function argument m is mutated
test/good/arg-copy-mutation-1.tact:5:9:
  4 |         // Bad cases:
> 5 |         m.set(3, 3);
              ^
  6 |         m.del(3);
Mutating function arguments has no effect outside the function due to call-by-value semantics
Help: Return the modified value or use the contract's state to avoid unnecessary mutations
See: https://nowarp.io/tools/misti/docs/detectors/ArgCopyMutation

[HIGH] ArgCopyMutation: Function argument m is mutated
test/good/arg-copy-mutation-1.tact:6:9:
  5 |         m.set(3, 3);
> 6 |         m.del(3);
              ^
  7 | 
Mutating function arguments has no effect outside the function due to call-by-value semantics
Help: Return the modified value or use the contract's state to avoid unnecessary mutations
See: https://nowarp.io/tools/misti/docs/detectors/ArgCopyMutation